src/Validation/Rule/NoUnusedVariablesRule.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace Digia\GraphQL\Validation\Rule;
+
+use Digia\GraphQL\Error\ValidationException;
+use Digia\GraphQL\Language\AST\Node\NodeInterface;
+use Digia\GraphQL\Language\AST\Node\OperationDefinitionNode;
+use Digia\GraphQL\Language\AST\Node\VariableDefinitionNode;
+use Digia\GraphQL\Language\AST\Node\VariableNode;
+use function Digia\GraphQL\Validation\unusedVariableMessage;
+
+/**
+ * No unused variables
+ *
+ * A GraphQL operation is only valid if all variables defined by an operation
+ * are used, either directly or within a spread fragment.
+ */
+class NoUnusedVariablesRule extends AbstractRule
+{
+    /**
+     * @var array|VariableDefinitionNode[]
+     */
+    protected $variableDefinitions;
+
+    /**
+     * @inheritdoc
+     */
+    public function enterNode(NodeInterface $node): ?NodeInterface
+    {
+        if ($node instanceof OperationDefinitionNode) {
+            $this->variableDefinitions = [];
```

Comment:
Is this correct? What if there are more than one operation definition nodes?

Reply:
Yes, the variables are checked for each OperationDefinitionNode separately.